### PR TITLE
Initial plumbing for safer name system

### PR DESCRIPTION
**Pull request by dougalm**
Handling names turns out to be one the trickier parts of writing a compiler. So far we've been following GHC's version of the Barendregt convention, "the rapier", described by Simon Peyton Jones and Simon Marlow in [Secrets of the GHC Inliner](https://www.microsoft.com/en-us/research/wp-content/uploads/2002/07/inline.pdf). It's elegant and it's fast. It's also stateless, which is great for caching and concurrency.

But it's *really* easy to screw up. We've messed it up [again](https://github.com/google-research/dex-lang/commit/e979cae84c9b0cd612bed1013cdecf71e8c0d917) and [again](https://github.com/google-research/dex-lang/commit/a6425c60a70b5db8871ea05185c30995d24bfbb8) and [again](https://github.com/google-research/dex-lang/commit/e979cae84c9b0cd612bed1013cdecf71e8c0d917) and [again](https://github.com/google-research/dex-lang/commit/c34ff0865306198aa9ed0c9ae1949325b6754dd7) and [again](https://github.com/google-research/dex-lang/commit/b96dbddba09bbd1e84f988da597bb350892c7fbd) and [again](https://github.com/google-research/dex-lang/commit/82c7edbde29c43e66eda8657de8853752709f11b) and [again](https://github.com/google-research/dex-lang/commit/c154995fa5eea42acef69d39b8247da5e455c8c1) .

This has become one of the biggest barriers to implementing new language features and onboarding new people. Worse, it makes us hesitate to use name-based indirection in places it would be helpful, like in de-duplicating data definitions and building long lived caches that persist across sessions.

We'll do a full write-up at some point, but the main idea is to keep using the same naming system as the Simons' rapier but enforce the discipline at compiler compile time using Haskell's type system. We add a phantom type parameter to every AST, representing the set of in-scope variables. Common mistakes -- like forgetting to extend the substitution as you go under a lambda, or hoisting subexpressions without checking for variables escaping their scope -- become compile time errors instead of the gnarly misnaming bugs we get today. Since the types are purely phantom, we don't affect the runtime performance at all, unlike schemes built around de Bruijn indices, like Ed Kmett's [bound](https://hackage.haskell.org/package/bound) library. ([Ed's write-up here](https://www.schoolofhaskell.com/user/edwardk/bound) was extremely helpful in understanding the problem and the solution space.)

We've been working on this change for a while (see [safer-names](https://github.com/google-research/dex-lang/tree/safer-names), [safer-names-3](https://github.com/google-research/dex-lang/tree/safer-names-3), [safer-names-4](https://github.com/google-research/dex-lang/tree/safer-names-4), [safer-names-5](https://github.com/google-research/dex-lang/tree/safer-names-5)) but it was hard to land as a monolithic change. Adam had the great idea last week to instead implement the new safe syntax alongside the old dangerous synax, with a bridge to convert between the two. That way we can switch over the compiler one pass at a time. This seems much more promising! This first PR just adds the data definitions and the bridge for converting between old and new syntax.

### CODE REVIEW

File: src/lib/SaferNames/Syntax.hs [hunk 1]
```diff
@@ -0,0 +1,585 @@
...
+      let s' = extendScope s b'
+      cont ext s' t' b'
+
+-- shorthand to makes instance implementations easier
+tce :: SubstE e => Monad m => Scope o -> CoreTraversal m i o -> e i -> m (e o)
+tce = traverseCoreE
+
+instance SubstB PlainBinder where
+  traverseCoreB = undefined
+  asBindingsFrag = undefined
```
Comment (apaszke) [inline src/lib/SaferNames/Syntax.hs:467]: There are many undefined instances here. Is that a blocker?

File: src/lib/SaferNames/Bridge.hs [hunk 2]
```diff
@@ -0,0 +1,220 @@
...
+  toSafeE name = UnsafeMakeName name
+  fromSafeE (UnsafeMakeName name) = name
+
+instance HasSafeVersionE D.Atom where
+  type SafeVersionE D.Atom = S.Atom
+
+  toSafeE atom = case atom of
+    D.Var v -> S.Var $ toSafeE v
+    D.Lam (D.Abs b (arr, body)) ->
+      S.Lam $ S.Abs (toSafeB b) (S.WithArrow (fmap toSafeE arr) (toSafeE body))
```
Comment (apaszke) [inline src/lib/SaferNames/Bridge.hs:53]: Do places like those assume that the dangerous representation already has no shadowing, or is that fine because this is not `FreshAbs`?

**Review by apaszke — COMMENTED** [review]
I think this mostly looks good. My main concern is the lack of docs, which has made it difficult to review this (I already paged out most of what we discussed and I expect this will happen in the future as well). It's also a bit annoying that it's not easy to get a diff between the current version of `Syntax` and the new one, but I'm just going to trust you on this one.

I'm also not entirely following the treatment of shadowing in the bridge. Do you just assume that it is all deshadowed and do unsafe casts into the new representation?

File: src/lib/Err.hs [hunk 3]
```diff
@@ -0,0 +1,128 @@
...
+ignoreExcept (Right x) = x
+
+assertEq :: (MonadError Err m, Show a, Pretty a, Eq a) => a -> a -> String -> m ()
+assertEq x y s = if x == y then return ()
+                           else throw CompilerErr msg
+  where msg = "assertion failure (" ++ s ++ "):\n"
+              ++ pprint x ++ " != " ++ pprint y ++ "\n"
+
+-- === small pretty-printing utils ===
+-- These are here instead of in PPrint.hs for import dependency reasons
```
[the remark this answers is not in the document]
Comment (dougalm) [reply src/lib/Err.hs:88]: I should say "import cycle reasons". `Err` is imported by `PPrint` via `Syntax`. But we want to use `pprint` in `Err` for `assertEq`, so we can't use the one defined in `PPrint`. (`assertEq` used to live in `PPrint` which gave confusing error messages.)

File: src/lib/SaferNames/LazyMap.hs [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+import qualified Data.Map.Strict as M
+
+-- This implements some of the Data.Map API, but also lets us apply fmap to it
+-- lazily without having to traverse every element. For example we might use it
+-- on a Name->Atom mapping that we want to temporarily view as a Name->Type
+-- mapping using `fmap getType`. Data.Map.Lazy almost has this property but (I
+-- think!) it still needs to touch every key, even if you only access a single
+-- element of the resulting map. (If that's not true then we should ditch this
+-- and use Data.Map.Lazy.)
+data LazyMap k a = LazyMap (S.Set k) (k -> a) (M.Map k a)
```
[the remark this answers is not in the document]
Comment (dougalm) [reply src/lib/SaferNames/LazyMap.hs:22]: The reason to carry the set is just to have `O(1)` access to the set of keys. We ask for the set of keys whenever we generate a fresh name based on the bindings env. I'll add some more comments.

File: src/lib/SaferNames/LazyMap.hs [hunk 5]
```diff
@@ -0,0 +1,82 @@
...
+
+lookupNoFailOption :: Ord k => k -> LazyMap k v -> v
+lookupNoFailOption k m = case lookup k m of
+  Just x -> x
+  Nothing -> error "failure is not an option"
+
+mapWithKey :: Ord k => (k -> a -> b) -> LazyMap k a -> LazyMap k b
+mapWithKey f lm = LazyMap (keysSet lm) (\k -> f k $ lookupNoFailOption k lm) mempty
+
+-- XXX: left-biased, like Data.Map, but unlike out Env-like structures.
```
[the remark this answers is not in the document]
Comment (dougalm) [reply src/lib/SaferNames/LazyMap.hs:64]: I make that typo  weirdly often...

File: src/lib/SaferNames/Name.hs [hunk 6]
```diff
@@ -0,0 +1,478 @@
...
+import qualified Data.Set        as S
+import Unsafe.Coerce
+import Data.Text.Prettyprint.Doc  hiding (nest)
+import GHC.Exts (Constraint)
+
+import qualified SaferNames.LazyMap as LM
+
+import qualified Env as D
+
+data S = (:=>:) S S
```
[the remark this answers is not in the document]
Comment (dougalm) [reply src/lib/SaferNames/Name.hs:45]: yes, coming up

File: src/lib/SaferNames/Name.hs [hunk 7]
```diff
@@ -0,0 +1,478 @@
...
+       | VoidScope
+
+-- TODO: we reuse the old `Name` to make use of the GlobalName name space while
+-- we're using both the old and new systems together.
+-- TODO: something like this instead:
+--    type Tag = T.Text
+--    data RawName = RawName Tag Int deriving (Show, Eq, Ord)
+type RawName = D.Name
+
+-- invariant: the raw name in `Name s` is contained in the list in the `ScopeVal s`
```
[the remark this answers is not in the document]
Comment (dougalm) [reply src/lib/SaferNames/Name.hs:57]: I renamed it `Scope` at some point but the comment is stale.

File: src/lib/SaferNames/Name.hs [hunk 8]
```diff
@@ -0,0 +1,478 @@
...
+
+newtype Env (n::S) (a:: *) = UnsafeMakeEnv (LM.LazyMap RawName a)
+                             deriving (Functor, Foldable, Traversable)
+type Scope n = Env n ()
+
+emptyEnv :: Env (n:=>:n) a
+emptyEnv = UnsafeMakeEnv mempty
+
+voidScopeEnv :: Env VoidScope a
+voidScopeEnv = UnsafeMakeEnv mempty
```
[the remark this answers is not in the document]
Comment (dougalm) [reply src/lib/SaferNames/Name.hs:69]: Just that they have different types. `Scope VoidScope` and `Scope (n:=>:n)` are both empty, but their types are not considered equal.

[the remark this answers is not in the document]
Comment (dougalm) [reply src/lib/SaferNames/Name.hs:69]: (It's not important that they have different types, it's just works out that way because of our representation. It would be better if we could have a single type for the empty set but it's harder to arrange.)

File: src/lib/SaferNames/Name.hs [hunk 9]
```diff
@@ -0,0 +1,478 @@
...
+
+class HasNamesB (b::B) where
+  traverseNamesB :: Monad m => Scope o -> RenameTraversal m i o
+                       -> b i i'
+                       -> m (FreshBinder b o (i:=>:i'))
+  boundScope :: b n l -> Scope (n:=>:l)
+
+-- slightly safer than raw `unsafeCoerce` because at least it checks the kind
+unsafeCoerceE :: forall (e::E) i o . e i -> e o
+unsafeCoerceE = unsafeCoerce
```
[the remark this answers is not in the document]
Comment (dougalm) [reply src/lib/SaferNames/Name.hs:143]: No, not yet. I'll try it in a follow-up patch.

File: src/lib/SaferNames/Name.hs [hunk 10]
```diff
@@ -0,0 +1,478 @@
...
+
+data PlainBinder (n::S) (l::S) where
+  UnsafeMakeBinder :: RawName -> PlainBinder n l
+  Ignore           :: PlainBinder n n
+
+type E = S -> *       -- expression-y things, covariant in the S param
+type B = S -> S -> *  -- binder-y things, covariant in the first param and
+                      -- contravariant in the second
+
+-- A `FreshExt n l` means that `l` extends `n` with only fresh names
```
[the remark this answers is not in the document]
Comment (dougalm) [reply src/lib/SaferNames/Name.hs:108]: The scopes represent a list of possibly-repeated raw names (though actually we represent them with sets because that's all we need -- maybe we should change that). A fresh `FreshExt n l` means that (1) the `Scope n` is a prefix of the `Scope l` and (2) The names in the diff (the list `Scope l` once you take off the common prefix) do not appear in that common prefix. Is that a clearer way to put it?

File: src/lib/SaferNames/Name.hs [hunk 11]
```diff
@@ -0,0 +1,478 @@
...
+data FreshExt (n::S) (l::S) = UnsafeMakeExt
+
+emptyFreshExt :: FreshExt n n
+emptyFreshExt = UnsafeMakeExt
+
+composeFreshExt :: FreshExt n h -> FreshExt h l -> FreshExt n l
+composeFreshExt _ _ = UnsafeMakeExt
+
+data NameTraversal (e::E) (m:: * -> *) (i::S) (o::S) where
```
[the remark this answers is not in the document]
Comment (dougalm) [reply src/lib/SaferNames/Name.hs:117]: We use it for the substitute-with-atoms traversal in Syntax.hs

File: src/lib/SaferNames/Name.hs [hunk 12]
```diff
@@ -0,0 +1,478 @@
...
+  NameTraversal :: (Name i -> m (e o))
+                -> RenameEnv (i:=>:i') o
```
[the remark this answers is not in the document]
Comment (dougalm) [reply src/lib/SaferNames/Name.hs:119]: docs coming right up!

File: src/lib/SaferNames/Name.hs [hunk 13]
```diff
@@ -0,0 +1,478 @@
...
+
+-- traverses free names, possibly renaming bound ones as needed
+class HasNamesE (e::E) where
+  traverseNamesE :: Monad m => Scope o -> RenameTraversal m i o -> e i -> m (e o)
+
+-- A fresh binder, hiding its new output scope and exposing a renaming
+-- env that can give you those names. Used mainly for the result of
+-- `traverseNamesB` and similar.
+data FreshBinder (b::B) (n::S) (i::S) where
+  FreshBinder :: FreshExt o o' -> b o o' -> RenameEnv i o'
```
[the remark this answers is not in the document]
Comment (dougalm) [reply src/lib/SaferNames/Name.hs:132]: Yes, renaming the variables is confusing, sorry. This one is a head scratcher and I don't love using it. But it seems like the most general thing for the result of `TraverseNamesB`. It might be easier to switch to the rank-2 continuation-oriented version of existentials for `TraverseNamesB`.

File: src/lib/SaferNames/Name.hs [hunk 14]
```diff
@@ -0,0 +1,478 @@
...
+              -> FreshBinder b o i
+
+class HasNamesB (b::B) where
+  traverseNamesB :: Monad m => Scope o -> RenameTraversal m i o
+                       -> b i i'
+                       -> m (FreshBinder b o (i:=>:i'))
```
[the remark this answers is not in the document]
Comment (dougalm) [reply src/lib/SaferNames/Name.hs:138]: I agree that it's hard to follow. It's actually the same pattern as our current [`renamingSubst`](https://github.com/google-research/dex-lang/blob/main/src/lib/Syntax.hs#L944) if that makes it any clearer. You get a fresh binder, along with a fragment of a renaming substitution that lets you map the names bound by the original binder to the new binder's inner name space.

File: src/lib/SaferNames/Name.hs [hunk 15]
```diff
@@ -0,0 +1,478 @@
...
+  boundScope :: b n l -> Scope (n:=>:l)
+
+-- slightly safer than raw `unsafeCoerce` because at least it checks the kind
+unsafeCoerceE :: forall (e::E) i o . e i -> e o
+unsafeCoerceE = unsafeCoerce
+
+unsafeCoerceB :: forall (b::B) n l n' l' . b n l -> b n' l'
+unsafeCoerceB = unsafeCoerce
+
+withFresh :: Scope n -> (forall l. FreshExt n l -> PlainBinder n l -> Name l -> a) -> a
```
[the remark this answers is not in the document]
Comment (dougalm) [reply src/lib/SaferNames/Name.hs:148]: It's just to avoid having to throw an error if you get the `Ignore` case.

File: src/lib/SaferNames/Name.hs [hunk 16]
```diff
@@ -0,0 +1,478 @@
...
+                  | i < bigInt  -> i + 1
+                  | otherwise   -> error "Ran out of numbers!"
+                _ -> 0
+    bigInt = (10::Int) ^ (9::Int)  -- TODO: consider a real sentinel value
+
+unitName :: Name UnitScope
+unitName = UnsafeMakeName "TheName"
+
+unitScope :: Scope UnitScope
+unitScope = UnsafeMakeEnv $ LM.singleton unitRawName ()
```
[the remark this answers is not in the document]
Comment (dougalm) [reply src/lib/SaferNames/Name.hs:168]: We use it in `foldMapNames`. It's for the same reason that you need a unit type if you want to, say, implement `fold :: Traversable f => Monoid a => f a -> a` using the writer monad.

File: src/lib/SaferNames/Name.hs [hunk 17]
```diff
@@ -0,0 +1,478 @@
...
+  where f' rawName = f $ UnsafeMakeName rawName
+
+envPairs :: Env (n:=>:l) a -> (PlainBinderList n l, [a])
+envPairs (UnsafeMakeEnv m) = do
+  let (names, vals) = unzip $ LM.assocs m
+  (unsafeMakePlainBinderList names, vals)
+
+unsafeMakePlainBinderList :: [RawName] -> PlainBinderList n l
+unsafeMakePlainBinderList names = case names of
+  [] -> unsafeCoerceB Empty
```
[the remark this answers is not in the document]
Comment (dougalm) [reply src/lib/SaferNames/Name.hs:182]: `Empty` can't have any type! The GADT constructor restricts it to have the same type parameter twice (which can be any type).

File: src/lib/SaferNames/Name.hs [hunk 18]
```diff
@@ -0,0 +1,478 @@
...
+  where UnsafeMakeName unitRawName = unitName
+
+envMapWithKey :: (Name i -> a -> b) -> Env i a -> Env i b
+envMapWithKey f (UnsafeMakeEnv m) = UnsafeMakeEnv $ LM.mapWithKey f' m
+  where f' rawName = f $ UnsafeMakeName rawName
+
+envPairs :: Env (n:=>:l) a -> (PlainBinderList n l, [a])
+envPairs (UnsafeMakeEnv m) = do
+  let (names, vals) = unzip $ LM.assocs m
+  (unsafeMakePlainBinderList names, vals)
```
[the remark this answers is not in the document]
Comment (dougalm) [reply src/lib/SaferNames/Name.hs:178]: Yes, an alternative would be something like,

```
evnPairs :: Env (n:=>:l) a -> Nest (AnnBinderP PlainBinder (LiftE a)) n l`
```
But unpacking nests is a pain because you have to worry about substituting the annotation at each step and so on. In this case, since the `a`s are actually all the same type (rather than each one having a slightly different scope) it seems more pain than it's worth. The lack of precision in the lengths seems like a small price to pay.

File: src/lib/SaferNames/Name.hs [hunk 19]
```diff
@@ -0,0 +1,478 @@
...
+  where t = newNameTraversal \name -> tell (f name) >> return unitName
+
+freeNames :: HasNamesE e => e n -> NameSet n
+freeNames e = foldMapNames S.singleton e
+
+-- should be equivalent to `\ext -> fmapNames (injectName ext)`
+injectNamesL :: HasNamesE e => FreshExt n l -> e n -> e l
+injectNamesL _ = unsafeCoerceE
+
+projectNamesL :: HasNamesE e => Scope n -> Scope (n:=>:l) -> e l -> Maybe (e n)
```
[the remark this answers is not in the document]
Comment (dougalm) [reply src/lib/SaferNames/Name.hs:205]: Good idea. If we just use the `Either (Name (n:=>:l))` naively then we'd just get the very first leaked name found. But maybe we should do `Either (Set (Name (n:=>:l)))`. Might be good for error reporting too. I'm hoping that being forced to use this will catch a lot of our leaked variable bugs and knowing their names will help fix them.

File: src/lib/SaferNames/Name.hs [hunk 20]
```diff
@@ -0,0 +1,478 @@
...
+type PlainBinderList = Nest PlainBinder
+
+infixl 7 :>
+data AnnBinderP (b::B) (ann ::E) (n::S) (l::S) =
+  (:>) (b n l) (ann n)
+  deriving (Show)
+
+infixl 7 :>>
+data AnnBinderListP (b::B) (ann::E) (n::S) (l::S) =
+  (:>>) (Nest b n l) [ann n]
```
[the remark this answers is not in the document]
Comment (dougalm) [reply src/lib/SaferNames/Name.hs:235]: yep. I'll add a comment.

File: src/lib/SaferNames/Name.hs [hunk 21]
```diff
@@ -0,0 +1,478 @@
...
+
+-- === environment variants ===
+
+type RenameEnv i o = Env i (Name o)
+
+-- covariant in `o`, contravariant in `i`.
+newtype EnvE (e::E) (i::S) (o::S) = EnvE { fromEnvE :: Env i (e o) }
+
+-- covariant in `n`
+newtype RecEnv (e::E) (n::S) = RecEnv { fromRecEnv :: Env n (e n) }
```
[the remark this answers is not in the document]
Comment (dougalm) [reply src/lib/SaferNames/Name.hs:245]: It's recursive because the payloads have the names in scope. For example:

```
-- nonrecursive let
data Let n l where Let :: Binder n l -> Expr n -> Let n l

--recursive let
data LetRec n l where LetRec :: Binder n l -> Expr l -> Let n l
```

File: src/lib/SaferNames/Name.hs [hunk 22]
```diff
@@ -0,0 +1,478 @@
...
+data ListE (e::E) (n::S) = ListE { fromListE :: [e n] }
+     deriving (Show, Eq)
+
+data LiftE (a:: *) (n::S) = LiftE { fromLiftE :: a }
+     deriving (Show, Eq)
+
+-- === various convenience utilities ===
+
+class HasNamesE e => AlphaEq (e::E) where
+  alphaEq :: Scope n -> e n -> e n -> Bool
```
[the remark this answers is not in the document]
Comment (dougalm) [reply src/lib/SaferNames/Name.hs:282]: Not yet. I tried to separate the changes cleanly but there are some fragments like these that refer to things that get used in follow-up changes. This will be used in the type checking pass which is the next PR.

File: src/lib/SaferNames/Name.hs [hunk 23]
```diff
@@ -0,0 +1,478 @@
...
+  pretty _ = "TODO"
+
+instance (PrettyB b) => Pretty (Nest b n l) where
+  pretty _ = "TODO"
+
+instance (PrettyB b, PrettyE e) => Pretty (Abs b e n) where
+  pretty _ = "TODO"
+
+instance (ShowB b, ShowE ann) => Show (AnnBinderListP b ann n l) where
+  show _ = "TODO"
```
[the remark this answers is not in the document]
Comment (dougalm) [reply src/lib/SaferNames/Name.hs:478]: Yeah, definitely. I'm starting to fill out pretty-printing instances in the next PR.

File: src/lib/SaferNames/Syntax.hs [hunk 24]
```diff
@@ -0,0 +1,585 @@
...
+-- TODO: get rid of annotations on variable occurrences
+data AnnVarP (ann::E) n = AnnVar (Name n) (ann n)
+                          deriving Show
+
+type Var = AnnVarP Type :: E
+type Binder     = AnnBinderP PlainBinder            Type  :: B
+type BinderList = AnnBinderP PlainBinderList (ListE Type) :: B
+
+data DataDef n where
+  DataDef :: RawName -> BinderList n l -> ListE DataConDef l -> DataDef n
```
[the remark this answers is not in the document]
Comment (dougalm) [reply src/lib/SaferNames/Syntax.hs:129]: I should add a comment. Getting this right was tricky. The `RawName` here is purely for pretty-printing. The actual binding name only appears in `UExpr` (safer-name version coming soon). By the time we're in Core, the actual name is just in the bindings.

File: src/lib/SaferNames/Name.hs [hunk 25]
```diff
@@ -0,0 +1,478 @@
...
+infixr 7 @>
+-- really, at most one name
+class HasNamesB b => BindsOneName (b::B) where
+  (@>) :: b n l -> a -> Env (n:=>:l) a
+
+infixr 7 @@>
+class HasNamesB b => BindsNameList (b::B) where
+  (@@>) :: b n l -> [a] -> Env (n:=>:l) a
+
+type EmptyNest (b::B) = Abs (Nest b) UnitE :: E
```
[the remark this answers is not in the document]
Comment (dougalm) [reply src/lib/SaferNames/Name.hs:293]: Agreed. I'll change it.

File: src/lib/SaferNames/Syntax.hs [hunk 26]
```diff
@@ -0,0 +1,585 @@
...
+type PiType  = Abs Binder (WithArrow Type)   :: E
+
+data WithArrow (e::E) (n::S) = WithArrow (Arrow n) (e n)
+                     deriving Show
+
+withoutArrow :: WithArrow e n -> e n
+withoutArrow (WithArrow _ x) = x
+
+justArrow :: WithArrow e n -> Arrow n
+justArrow (WithArrow arr _) = arr
```
[the remark this answers is not in the document]
Comment (dougalm) [reply src/lib/SaferNames/Syntax.hs:149]: Can you do that with GADTs? If so, I should do more of that!

File: src/lib/SaferNames/Syntax.hs [hunk 27]
```diff
@@ -0,0 +1,585 @@
...
+
+data Module n where
+  Module :: IRVariant
+         -> Nest Decl n l
+         -> BindingsFrag l l'
+         -> SourceNameMap l'
+         -> Module n
+
+type WithBindings = Abs BindingsFrag
+type EvaluatedModule = WithBindings SourceNameMap  :: E
```
[the remark this answers is not in the document]
Comment (dougalm) [reply src/lib/SaferNames/Syntax.hs:177]: Right, that's why I started adding kind declarations (though missing one here). Will do.

File: src/lib/SaferNames/Syntax.hs [hunk 28]
```diff
@@ -0,0 +1,585 @@
...
+ | DataDefName (DataDef n)
+ | ClassName -- TODO
+ | MethodName  (NamedDataDef n) Int  -- TODO: just have a class name and a label
+ | DataConName (NamedDataDef n) Int
+ | TypeConName (NamedDataDef n)
+ -- TODO: avoid this case by having the inference monad use an alternative
+ -- version of BinderInfo
+ | InferenceName
+
+data TypedBinderInfo n = TypedBinderInfo (Maybe (Type n)) (BinderInfo n)
```
[the remark this answers is not in the document]
Comment (dougalm) [reply src/lib/SaferNames/Syntax.hs:222]: It's because things there are bindings for things that aren't atom-like, like the `DataDefName`, that don't have names. There are other solutions, like carrying the "static info payload" type on the name itself. But that's for the next rewrite, I think.

[the remark this answers is not in the document]
Comment (dougalm) [reply src/lib/SaferNames/Syntax.hs:222]: (I'll add a note)

File: src/lib/SaferNames/Syntax.hs (hunk 1 above)
Comment (dougalm) [reply src/lib/SaferNames/Syntax.hs:467]: Follow-up PR!

File: src/lib/SaferNames/Bridge.hs (hunk 2 above)
Comment (dougalm) [reply src/lib/SaferNames/Bridge.hs:53]: That's right. Ordinary `Abs` doesn't have a freshness guarantee. (`FreshAbs` explicitly carries a `FreshExt`, which is the only way you get a freshness guarantee.)

File: src/lib/SaferNames/Syntax.hs [hunk 29]
```diff
@@ -0,0 +1,585 @@
...
+-- === traversals with atom substitutions ===
+
+data SubstVal e n = SubstVal (e n) | Rename (Name n)
+
+-- TODO: make this just `Env (Atom o) i` if we remove type annotations from
+-- variable occurrences
+type AtomSubstVal = SubstVal Atom
+type SubstEnv i o = Env i (AtomSubstVal o)
+
+type CoreTraversal = NameTraversal AtomSubstVal
```
[the remark this answers is not in the document]
Comment (dougalm) [reply src/lib/SaferNames/Syntax.hs:236]: Not sure what you mean. The CoreTraversal is useful for things like beta reduction, where mere renaming isn't sufficient. But we can implement `HasNamesE` in terms of `SubstE`, if that's what you mean.

File: src/lib/Err.hs (hunk 3 above)
[the remark this answers is not in the document]
Comment (apaszke) [reply src/lib/Err.hs]: I'd just go with `.hs-boot` and put things where they belong. But 🤷‍♂️

File: src/lib/SaferNames/Syntax.hs (hunk 29 above)
[the remark this answers is not in the document]
Comment (apaszke) [reply src/lib/SaferNames/Syntax.hs:233]: We do have a traversal for rewriting terms in Builder (look for `TraversalDef`). This is called similarly, but seems much more restricted.

[the remark this answers is not in the document]
Comment (dougalm) [reply src/lib/SaferNames/Syntax.hs:233]: Ah, I misread your comment. I thought you were comparing the taversal in Name.hs with the core-specific one in Syntax.hs. Yes, this is strictly weaker than the one in builder. Maybe we can de-dup at some stage.
